finetune/util/table_labeler.py

```diff
@@ -0,0 +1,331 @@
+            "table_context": table_context,
+            "table_doc_i": table_doc_i,
+            "doc_text": doc_text,
+        }
+
+    def resolve_preds(
+        self, table_preds, text_preds, table_chunks, document_text, table_doc_i
+    ):
+        output_preds = []
+        table_pred_chunks = [[] for _ in document_text]
+        for p, c, i in zip(table_preds, table_chunks, table_doc_i):
+            table_pred_chunks[i].append((p, c))
+
+        for (i, (text_preds_i, document_text_i, doc_pred_chunks)) in enumerate(
+            zip(text_preds, document_text, table_pred_chunks)
+        ):
+            output_doc_preds = []
+            for p, c in doc_pred_chunks:
+                output_doc_preds += self.fix_spans(
+                    p,
+                    chunk=c,
+                    output_space_text=document_text_i,
+                    input_space="table",
+                    output_space="document",
+                )
+            if self.drop_table_from_text_preds:
+                output_doc_preds += self.remove_table_labels(
+                    text_preds_i, [d[1] for d in doc_pred_chunks]
+                )
+            else:
+                output_doc_preds += text_preds_i
+            output_preds.append(sorted(output_doc_preds, key=lambda x: x["start"]))
+        return output_preds
+
+
+@functools.lru_cache
+def get_etl_from_file(model_file_path):
+    return SequenceLabeler.load(model_file_path, key="etl")
+
+
+class TableLabeler:
+    def __init__(
+        self,
+        table_model_config: t.Dict[str, t.Any] = None,
+        normal_model_config: t.Dict[str, t.Any] = None,
+        drop_table_from_text_labels: bool = False,
+        drop_table_from_text_preds: bool = True,
+    ):
+        self.etl = TableETL(
+            drop_table_from_text_labels=drop_table_from_text_labels,
+            drop_table_from_text_preds=drop_table_from_text_preds,
+        )
+        self.table_model = SequenceLabeler(
+            base_model=TableRoBERTa, **(table_model_config or {})
+        )
+        self.text_model = SequenceLabeler(
+            base_model=RoBERTa, **(normal_model_config or {})
+        )
+
+    def fit(
+        self,
+        text: t.List[str],
+        tables: t.List[t.List[t.Dict[str, t.Any]]],
+        labels: t.List[t.List[t.Dict[str, t.Any]]],
+    ):
+        model_inputs = self.etl.get_table_text_chunks_and_context(
+            text=text, tables=tables, labels=labels
+        )
+        self.table_model.fit(
+            model_inputs["table_text"],
+            model_inputs["table_labels"],
+            context=model_inputs["table_context"],
+        )
+        self.text_model.fit(model_inputs["doc_text"], model_inputs["doc_labels"])
+
+    def save(self, path: str):
+        SequenceLabeler.save_multiple(
+            path,
+            models={
+                "table": self.table_model,
+                "text": self.text_model,
+                "etl": self.etl,
```

Comment:
Does the ETL need to be serialized because you need the "self.drop_table_from_text_labels" and "self.drop_table_from_text_preds" settings?